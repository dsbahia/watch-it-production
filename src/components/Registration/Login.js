import { React, useState } from "react";
import {
  signInWithEmailAndPassword,
  sendEmailVerification,
} from "firebase/auth";
import { toast } from "react-hot-toast";
import { Link, useNavigate } from "react-router-dom";
import { auth } from "./firebase";
import { useAuthValue } from "./AuthContext";
import WatchItLogo from "../WatchItLogo";
import "../../styles/forms.css";

function Login() {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const { setTimeActive } = useAuthValue();
  const navigate = useNavigate();

  const login = async (e) => {
    e.preventDefault();

    try {
      await signInWithEmailAndPassword(auth, email, password);

      if (!auth.currentUser.emailVerified) {
        await sendEmailVerification(auth.currentUser);
        setTimeActive(true);
        navigate("/verify-email");
      } else {
        navigate("/");
      }
    } catch (err) {
      setError("");
      if (err.code === "auth/invalid-login-credentials") {
        toast.error("Incorrect email or password. Please try again.");
      } else {
        toast.error("An error occurred. Please try again later.");
      }
    }
  };

  return (
    <div className="auth-container">
      <div className="auth-form">
        <WatchItLogo />
        <h1>Log in</h1>
        {error && <div className="auth-error">{error}</div>}
        <form onSubmit={login} name="login_form">
          <input
            type="email"
            value={email}
            required
            placeholder="Enter your email"
            onChange={(e) => setEmail(e.target.value)}
            className="auth-input"
          />

          <input
            type="password"
            value={password}
            required
            placeholder="Enter your password"
            onChange={(e) => setPassword(e.target.value)}
            className="auth-input"
          />

          <button type="submit" className="auth-button">
            Login
          </button>
        </form>
        <span>
          Don&apos;t have an account?
          <Link to="/register">Create one here</Link>
        </span>
      </div>
    </div>
  );
}

export default Login;
