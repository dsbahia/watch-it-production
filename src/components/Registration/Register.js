import { React, useState } from "react";
import { useNavigate, Link } from "react-router-dom";
import { toast } from "react-hot-toast";
import {
  createUserWithEmailAndPassword,
  sendEmailVerification,
} from "firebase/auth";
import { auth } from "./firebase";
import { useAuthValue } from "./AuthContext";
import WatchItLogo from "../WatchItLogo";
import "../../styles/forms.css";

function Register() {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const { setTimeActive } = useAuthValue();

  const navigate = useNavigate();

  const showPasswordValidationError = (errorMessage) => {
    toast.error(errorMessage);
  };

  const validatePassword = () => {
    if (password !== confirmPassword) {
      showPasswordValidationError("Passwords do not match");
      return false;
    }
    return true;
  };

  const register = async (e) => {
    e.preventDefault();
    toast.dismiss();

    if (validatePassword()) {
      try {
        await createUserWithEmailAndPassword(auth, email, password);
        await sendEmailVerification(auth.currentUser);
        setTimeActive(true);
        navigate("/verify-email");
      } catch (err) {
        const errorMap = {
          "auth/email-already-in-use": "Email is already in use.",
          "auth/invalid-email": "Invalid email address.",
          "auth/weak-password":
            "Password is too weak. Please choose a stronger password.",
        };

        const errorMessage =
          errorMap[err.code] || "An error occurred. Please try again later.";
        toast.error(errorMessage);
      }

      setEmail("");
      setPassword("");
      setConfirmPassword("");
    }
  };

  return (
    <div className="auth-container">
      <div className="auth-form">
        <WatchItLogo />
        <h1>Register</h1>
        <form onSubmit={register} name="registration_form">
          <input
            type="email"
            value={email}
            placeholder="Enter your email"
            required
            onChange={(e) => setEmail(e.target.value)}
            className="auth-input"
          />

          <input
            type="password"
            value={password}
            required
            placeholder="Enter your password"
            onChange={(e) => setPassword(e.target.value)}
            className="auth-input"
          />

          <input
            type="password"
            value={confirmPassword}
            required
            placeholder="Confirm password"
            onChange={(e) => setConfirmPassword(e.target.value)}
            className="auth-input"
          />

          <button type="submit" className="auth-button">
            Register
          </button>
        </form>
        <span>
          Already have an account?
          <Link to="/login">Login</Link>
        </span>
      </div>
    </div>
  );
}

export default Register;
