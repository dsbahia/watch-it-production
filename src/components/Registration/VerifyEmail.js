import React, { useState, useEffect } from "react";
import { toast } from "react-hot-toast";
import { useNavigate } from "react-router-dom";
import { sendEmailVerification } from "firebase/auth";
import { auth } from "./firebase";
import { useAuthValue } from "./AuthContext";
import "../../styles/verifyEmail.css";

function VerifyEmail() {
  const { currentUser } = useAuthValue();
  const [time, setTime] = useState(60);
  const { timeActive, setTimeActive } = useAuthValue();
  const navigate = useNavigate();

  useEffect(() => {
    const interval = setInterval(async () => {
      if (currentUser) {
        try {
          await currentUser.reload();
          if (currentUser.emailVerified) {
            clearInterval(interval);
            navigate("/");
          }
        } catch (err) {
          toast.error("An error occurred. Please try again later.");
        }
      }
    }, 1000);
  }, [navigate, currentUser]);

  useEffect(() => {
    let timeInterval = null;
    if (timeActive && time !== 0) {
      timeInterval = setInterval(() => {
        setTime((prevTime) => prevTime - 1);
      }, 1000);
    } else if (time === 0) {
      setTimeActive(false);
      setTime(60);
      clearInterval(timeInterval);
    }
    return () => clearInterval(timeInterval);
  }, [timeActive, time, setTimeActive]);

  const resendEmailVerification = async () => {
    if (auth.currentUser) {
      try {
        await sendEmailVerification(auth.currentUser);
        setTimeActive(true);
      } catch (err) {
        toast.error("An error occurred. Please try again later.");
      }
    }
  };

  return (
    <div className="verify-email-container">
      <div className="verify-email-contents">
        <h1>Verify your Email Address</h1>
        <p>
          A Verification email has been sent to:{" "}
          <strong>{currentUser && currentUser.email}</strong>
        </p>
        <p>Follow the instructions in the email to verify your account.</p>
        <button
          onClick={resendEmailVerification}
          type="button"
          disabled={timeActive}
        >
          Resend Email {timeActive && `(${time})`}
        </button>
      </div>
    </div>
  );
}

export default VerifyEmail;
