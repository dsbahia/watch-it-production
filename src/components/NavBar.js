import React from "react";
import { Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faUser,
  faSignInAlt,
  faSearch,
} from "@fortawesome/free-solid-svg-icons";
import MovieDropDown from "./DropDowns/MovieDropDown";
import TvDropDown from "./DropDowns/TvDropDown";
import { useAuthValue } from "./Registration/AuthContext";
import "../styles/navbar.css";

function NavBar() {
  const { currentUser } = useAuthValue();
  return (
    <div className="navbar">
      <div className="navbar-items">
        <Link className="nav-link" to="/">
          <FontAwesomeIcon icon={faSearch} className="navbar-icon" /> Search
        </Link>
        {!currentUser ? (
          <>
            <Link className="nav-link" to="/register">
              <FontAwesomeIcon icon={faUser} className="navbar-icon" /> Register
            </Link>
            <Link className="nav-link" to="/login">
              <FontAwesomeIcon icon={faSignInAlt} className="navbar-icon" />{" "}
              Login
            </Link>
          </>
        ) : null}

        {currentUser ? (
          <Link className="nav-link" to="/profile">
            <FontAwesomeIcon icon={faUser} className="navbar-icon" /> Profile
          </Link>
        ) : null}

        <MovieDropDown />
        <TvDropDown />
      </div>
    </div>
  );
}

export default NavBar;
