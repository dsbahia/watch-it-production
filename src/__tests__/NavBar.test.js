import React from "react";
import { MemoryRouter, Link } from "react-router-dom";
import { render } from "@testing-library/react";
import { useAuthValue } from "../components/Registration/AuthContext";
import NavBar from "../components/NavBar";
import "@testing-library/jest-dom/extend-expect";

jest.mock("../components/Registration/AuthContext", () => ({
  useAuthValue: jest.fn(),
}));

describe("NavBar", () => {
  it("renders correctly", () => {
    useAuthValue.mockReturnValue({
      currentUser: null,
    });

    const { asFragment } = render(
      <MemoryRouter>
        <NavBar />
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a link with the correct className", () => {
    useAuthValue.mockReturnValue({
      currentUser: null,
    });
    const { container } = render(
      <MemoryRouter>
        <NavBar />
      </MemoryRouter>,
    );
    const linkElement = container.querySelector("a");

    expect(linkElement).toHaveClass("nav-link");
  });

  it("renders the link with the correct route", () => {
    useAuthValue.mockReturnValue({
      currentUser: null,
    });
    const { container } = render(
      <MemoryRouter>
        <Link to="/" className="item-upcoming">
          Upcoming
        </Link>
      </MemoryRouter>,
    );
    const linkElement = container.querySelector("a");

    expect(linkElement).toHaveAttribute("href", "/");
  });
});
