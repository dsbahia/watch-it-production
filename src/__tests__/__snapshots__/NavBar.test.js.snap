// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar renders correctly 1`] = `
<DocumentFragment>
  <div
    class="navbar"
  >
    <div
      class="navbar-items"
    >
      <a
        class="nav-link"
        href="/"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-magnifying-glass navbar-icon"
          data-icon="magnifying-glass"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
            fill="currentColor"
          />
        </svg>
         Search
      </a>
      <a
        class="nav-link"
        href="/register"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-user navbar-icon"
          data-icon="user"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
            fill="currentColor"
          />
        </svg>
         Register
      </a>
      <a
        class="nav-link"
        href="/login"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-right-to-bracket navbar-icon"
          data-icon="right-to-bracket"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M217.9 105.9L340.7 228.7c7.2 7.2 11.3 17.1 11.3 27.3s-4.1 20.1-11.3 27.3L217.9 406.1c-6.4 6.4-15 9.9-24 9.9c-18.7 0-33.9-15.2-33.9-33.9l0-62.1L32 320c-17.7 0-32-14.3-32-32l0-64c0-17.7 14.3-32 32-32l128 0 0-62.1c0-18.7 15.2-33.9 33.9-33.9c9 0 17.6 3.6 24 9.9zM352 416l64 0c17.7 0 32-14.3 32-32l0-256c0-17.7-14.3-32-32-32l-64 0c-17.7 0-32-14.3-32-32s14.3-32 32-32l64 0c53 0 96 43 96 96l0 256c0 53-43 96-96 96l-64 0c-17.7 0-32-14.3-32-32s14.3-32 32-32z"
            fill="currentColor"
          />
        </svg>
         Login
      </a>
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          class="dropdown-toggle btn btn-primary"
          id="item-movie-dropdown"
          type="button"
        >
          Movies
        </button>
      </div>
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          class="dropdown-toggle btn btn-primary"
          id="item-tv-dropdown"
          type="button"
        >
          Tv Shows
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
